Add JSON format for EXPLAIN with type LOGICAL and DISTRIBUTED

chyikwei opened the pull request:
Add Json format for `EXPLAIN` with type `LOGICAL` and `DISTRIBUTED`. (Issue #11189)

- for both Logical and distributed type, I use `JsonRenderedNode` in  JsonRenderer ([link](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/JsonRenderer.java#L57-L111))

- Distributed plan is using sorted map base on [QueryMonitor](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java#L328). (Also moved `JsonPlanFragment` from QueryMonitor to JsonRenderer.)

- sample output in [this gist](https://gist.github.com/chyikwei/10fda9fe1bbb31f16ecfac5ff9b59938)

```
== RELEASE NOTES ==

General Changes
* Add JSON format for `EXPLAIN` with type `LOGICAL` and `DISTRIBUTED`.

chyikwei commented:
@rschlussel / @rongrong:
I have a question about json format for `DISTRIBUTED` plan:
- In [textDistributedPlan](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java#L243-L245), I see each `PlanFragment` is printed one by one as a list. 
- But in [QueryMonitor](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java#L328),  fragments are kept in a sorted map with `FragmentId` as the key of each fragment.

For now I followed the TEXT format and stored it as a list, but want to check which is the preferred format for it? (a list of a map?)

rschlussel commented:
Good question! I think a sorted map is probably better, so that all the things that produce a json plan do so in the same format (json map). For the text plan the internal format for how it's stored isn't really relevant, since it's more about the text that gets produced.

chyikwei commented:
updated the distributed  format. sample output looks like this:
```
presto> explain (type distributed, format json) select * from example.example.numbers;
                                                                                                               Query Plan
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {
   "0" : {
     "plan" : {
   "id" : "5",
   "name" : "Output",
   "identifier" : "[text, value]",
   "details" : "",
   "children" : [ {
     "id" : "63",
     "name" : "RemoteSource",
     "identifier" : "[1]",
     "details" : "",
     "children" : [ ],
     "remoteSources" : [ "1" ]
   } ],
   "remoteSources" : [ ]
 }
   },
   "1" : {
     "plan" : {
   "id" : "0",
   "name" : "TableScan",
   "identifier" : "[TableHandle {connectorId='example', connectorHandle='example:example:numbers', layout='Optional[example:example:numbers]'}]",
   "details" : "text := ExampleColumnHandle{connectorId=example, columnName=text, columnType=varchar, ordinalPosition=0}\nvalue := ExampleColumnHandle{connectorId=example, columnName=value, columnType=bigint, ord
   "children" : [ ],
   "remoteSources" : [ ]
 }
   }
 }
(1 row)
```

(Also I cannot reproduce travis CI test fail on my local. Will see if rebased can fix it.)

chyikwei commented:
@rschlussel can you help review this PR?

rschlussel commented:
@chyikwei I'll be taking off tomorrow and Friday and a couple days next week to celebrate Passover.  I'm happy to review the PR, but not sure exactly how soon that will be.

chyikwei commented:
@rschlussel np. take your time and thanks for your help!

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -247,6 +249,53 @@ public static String textDistributedPlan(SubPlan plan, FunctionManager functionM
+        return new PlanPrinter(plan, types, stageExecutionStrategy, functionManager, estimatedStatsAndCosts, session, stats).toJson();
+    }
+
+    public static String jsonLogicalPlan(
```
nit: put this method before the previous one.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java:
```diff
@@ -160,10 +160,17 @@ public String getJsonPlan(Session session, Statement statement, Type planType, L
                 return textIOPlan(plan.getRoot(), metadata, session);
+            case LOGICAL:
+                plan = getLogicalPlan(session, statement, parameters, warningCollector);
+                return PlanPrinter.jsonLogicalPlan(plan.getRoot(), plan.getTypes(), metadata.getFunctionManager(), plan.getStatsAndCosts(), session);
```
nit: static import

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java:
```diff
@@ -160,10 +160,17 @@ public String getJsonPlan(Session session, Statement statement, Type planType, L
+                return PlanPrinter.jsonLogicalPlan(plan.getRoot(), plan.getTypes(), metadata.getFunctionManager(), plan.getStatsAndCosts(), session);
+            case DISTRIBUTED:
+                SubPlan subPlan = getDistributedPlan(session, statement, parameters, warningCollector);
+                return PlanPrinter.jsonDistributedPlan(subPlan);
```
static import

chyikwei commented:
@rschlussel thanks for review this and sorry for the late reply. (didn't notice you left comments)
I addressed the comments and please take a look. thanks!

rschlussel commented:
Merged, thanks!
